File: 백준/Gold/14502. 연구소/연구소.cc
#include <iostream>
#include <queue>
using namespace std;

#define P pair<int, int>
#define X first
#define Y second

/**
# boj14502 연구소

## 설계
1. 벽 3개 경우의 수
벽 3개 세우기
완탐 최대 64^3 = 262144

2. 바이러스 퍼뜨리기 
bfs로 2인 곳에서부터 0으로 상하좌우로 다 퍼뜨린 뒤,
안전영역 갯수 세기
O(V+E) = O(4NM) = O(4*64)

=> 가능
*/

int n,m;

const int MX = 12;
int board[MX][MX];
int board2[MX][MX];
bool ch[MX][MX];
bool ch2[MX][MX];

P arr[MX];

int dx[4] = {1,0,-1,0};
int dy[4] = {0,1,0,-1};

int mx = 0;

void bfs(int sx, int sy){
    queue<P> q;
    
    q.push({sx,sy});
    ch2[sx][sy] = 1;
    
    while(!q.empty()){
        int sz = q.size();
        
        while(sz--){
            int x = q.front().X;
            int y = q.front().Y;
            // board2[x][y] = 2;
            q.pop();
            
            for(int i = 0; i <4; i++){
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if(nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                if(board2[nx][ny] != 0 || ch2[nx][ny]) continue;
                q.push({nx,ny});
                ch2[nx][ny] = 1;
                board2[nx][ny] = 2;
            }
        }
    }
    

}

void recur(int cur){
    if(cur == 3){
        // 초기화
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
              ch2[i][j] = 0;
              board2[i][j] = board[i][j];
            }
        }
        // 벽세우기
        for(int i = 0; i < 3; i++){
            board2[arr[i].X][arr[i].Y] = 1;
        }
        
        // // 바이러스 퍼뜨리기
        for(int i = 0; i< n; i++){
            for(int j = 0; j< m; j++){
                if(ch2[i][j] || board2[i][j] != 2) continue;
                bfs(i,j);
            }
        }
        
        int cnt = 0;    
        for(int i = 0; i< n; i++){
            for(int j = 0; j< m; j++){
                if(board2[i][j] == 0) cnt++;
            }
        }
        
        mx = max(cnt, mx);
        return;
    }
    
    for(int i = 0; i< n; i++){
        for(int j = 0; j< m; j++){
            if(ch[i][j] || board[i][j] != 0) continue;
            arr[cur] = {i,j};
            ch[i][j] = 1;
            recur(cur+1);
            ch[i][j] = 0;
        }
    }
}

int main()
{
    cin >> n >> m;
    for(int i = 0; i< n; i++){
        for(int j = 0; j< m; j++){
            cin >> board[i][j];
        }
    }
    
    recur(0);

    cout << mx;
    return 0;
}

File: 백준/Gold/14502. 연구소/연구소.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	static int N, M;
	static int dx[] = new int[] { 1, 0, -1, 0 };
	static int dy[] = new int[] { 0, -1, 0, 1 };
	static int tmp[] = new int[3];
	static boolean ch[][] = new boolean[10][10];

	static int ori[][] = new int[10][10];
	static int map[][] = new int[10][10];
	static int mx;

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < M; j++) {
				ori[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		recur(0, 0);
		System.out.println(mx);
	}

	static void recur(int cur, int st) {
		if (cur == 3) {
			solve();
			return;
		}

		for (int i = st; i < N * M; i++) {
			int x = i / M;
			int y = i % M;
			if (ori[x][y] != 0)
				continue;
			tmp[cur] = i;
			recur(cur + 1, i + 1);
		}
	}

	static void solve() {
		ch = new boolean[10][10];
		copy();
		for (int i : tmp) {
			int x = i / M;
			int y = i % M;

			map[x][y] = 1;
		}

		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				if (map[i][j] != 2 || ch[i][j])
					continue;
				bfs(i, j);
			}
		}

		int cnt = 0;
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				if (map[i][j] == 0)
					cnt++;
			}
		}
		mx = Math.max(mx, cnt);
	}

	static void bfs(int sx, int sy) {
		Queue<Integer[]> q = new ArrayDeque<>();

		q.add(new Integer[] { sx, sy });
		ch[sx][sy] = true;

		while (!q.isEmpty()) {
			int sz = q.size();
			while (sz-- > 0) {
				Integer[] cur = q.poll();
				int x = cur[0];
				int y = cur[1];
				map[x][y] = 2;

				for (int dir = 0; dir < 4; dir++) {
					int nx = x + dx[dir];
					int ny = y + dy[dir];
					if (nx < 0 || nx >= N || ny < 0 || ny >= M)
						continue;
					if (map[nx][ny] != 0 || ch[nx][ny])
						continue;
					q.add(new Integer[] { nx, ny });
					ch[nx][ny] = true;
				}
			}
		}
	}

	static void copy() {
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				map[i][j] = ori[i][j];
			}
		}
	}

}